Add todo details screen for the iOS sample app

sgallese opened the pull request:
https://github.com/arkivanov/MVIKotlin/issues/18

Still needs some refactoring and cleanup before merge.

Please add any questions/comments/suggestions, very open to improvements :)

sgallese commented on sample/todo-app-ios/todo-app-ios/list/TodoListViewImpl.swift:
```diff
@@ -0,0 +1,39 @@
+    override func render(model: TodoListViewModel) {
+        self.model = model
+        
+        DiffKt.diff { (diffBuilder: DiffBuilder) in
```
I tried to copy logic from Android here. I can remove this render diff code if not needed.

sgallese commented on sample/todo-app-ios/todo-app-ios/details/TodoDetails.swift:
```diff
@@ -0,0 +1,72 @@
+    @State var lifecycle = LifecycleRegistry()
+    
+    var body: some View {
+        let text = Binding<String>(
```
Hmm, maybe these text and boolean bindings should be moved into TodoDetailsViewImpl

sgallese replied:
Didn't end up moving to the TodoDetailsViewImpl... looking at code examples, looks like folks simply inline the binding to where it is used, so I did that with 9ea2c28

sgallese commented on sample/todo-app-ios/todo-app-ios/details/TodoDetails.swift:
```diff
@@ -0,0 +1,72 @@
+            }
+            
+        }.onAppear() {
+            let controller = TodoDetailsReaktiveController(
```
This feels like a lot of boilerplate for each swift view lifecycle.
Maybe the project needs a Scene delegate lifecycle and SwiftUI view lifecycle helper (similar to Android's Arch Lifecycle implementation for this project).

sgallese commented on sample/todo-app-ios/todo-app-ios/details/TodoDetails.swift:
```diff
@@ -0,0 +1,72 @@
+            set: { bool in self.detailsView.dispatch(event: TodoDetailsViewEvent.DoneClicked()) }
+        )
+        
+        return VStack(alignment: .leading, spacing: 20) {
```
Spacing is  ugly, can spend 5 mins trying to make prettier

sgallese replied:
This was a challenge, details of success/failure here: https://github.com/arkivanov/MVIKotlin/issues/18#issuecomment-613812082

sgallese commented:
Interesting the iOS build failed on CI.
Investigating. Looks OK for me locally.
"ERROR: Exception while analyzing expression at (11,9) in /Users/runner/runners/2.168.0/work/MVIKotlin/MVIKotlin/mvikotlin/src/commonMain/kotlin/com/arkivanov/mvikotlin/core/lifecycle/LifecycleRegistry.kt"

arkivanov commented:
> Interesting the iOS build failed on CI.
> The app compiles on iOS for me locally.
> 
> Running ./gradlew clean :mvikotlin:compileKotlinIosX64 locally also has no issue.
> 
> Maybe the Github Action needs to be re-run?
> 
> "ERROR: Exception while analyzing expression at (11,9) in /Users/runner/runners/2.168.0/work/MVIKotlin/MVIKotlin/mvikotlin/src/commonMain/kotlin/com/arkivanov/mvikotlin/core/lifecycle/LifecycleRegistry.kt"

Looks like now it's ok

arkivanov commented:
I will review this PR as soon as I have time. Thanks!

arkivanov commented on sample/todo-app-ios/todo-app-ios/details/TodoDetails.swift:
```diff
@@ -0,0 +1,72 @@
+    @EnvironmentObject var controllerDeps: ControllerDeps
+    
+    @Environment(\.presentationMode) var presentation
+    @State var lifecycle = LifecycleRegistry()
```
Do we need `@State` here?

sgallese replied:
When I didn't use state here, I would encounter lifecycle errors such as
`Uncaught Kotlin exception: kotlin.IllegalStateException: Expected lifecycle state RESUMED, actual INITIALIZED`

Let me see if it's still needed once I set up a parent view per https://github.com/arkivanov/MVIKotlin/pull/33#discussion_r410713086

arkivanov commented on sample/todo-app-ios/todo-app-ios/details/TodoDetails.swift:
```diff
@@ -0,0 +1,72 @@
+    var id: String
+    
+    @ObservedObject var detailsView = TodoDetailsViewImpl()
+    @EnvironmentObject var controllerDeps: ControllerDeps
```
I don't like that `TodoDetails` view has that many responsibilities compared to the `TodoList` view. Can we unify them somehow? Let's think about it. Ideally controllers and their deps should leave somewhere outside. But I'm not sure.

sgallese replied:
Yes, we can make a parent view for TodoDetails that wires this up.
Essentially that is what the SceneDelegate is acting as for TodoList/TodoAdd.

arkivanov commented on sample/todo-app-ios/todo-app-ios/details/TodoDetails.swift:
```diff
@@ -0,0 +1,72 @@
+            .navigationBarTitle("Details")
+            .navigationBarItems(trailing: Button(action: {
+                self.detailsView.dispatch(event: TodoDetailsViewEvent.DeleteClicked())
+                self.presentation.wrappedValue.dismiss()
```
Could you please explain briefly what is it and how `EnvironmentObject` works?

sgallese replied:
To do this without EnvironmentObject, we would pass ControllerDeps (ObservableObject) to TodoList.
Then when TodoList creates TodoDetails, we would pass ControllerDeps to TodoDetails.

EnvironmentObject allows you skip explicit passing of ObservableObjects to nested child views.

I'm down with either approach if you have a preference.

arkivanov commented:
Also I can't compile the Swift project because of strange errors:
![image](https://user-images.githubusercontent.com/26204457/79616022-f263e100-80fb-11ea-8012-7c212d0b0fe8.png)
![image](https://user-images.githubusercontent.com/26204457/79616087-10314600-80fc-11ea-9daf-e898ab3274e0.png)

For some reason it does not like passing `@Published` and `@ObservedObject` objects to Kotlin classes. But if I e.g. specify types explicitly, then it works:
```swift
class ControllerDeps: ObservableObject {
    @Published var storeFactory: StoreFactory = LoggingStoreFactory(
        delegate: TimeTravelStoreFactory(fallback: DefaultStoreFactory()),
        logger: DefaultLogger(),
        mode: LoggingMode.full,
        eventTypes: StoreEventType.Companion().ALL
    )
    
    @Published var database: TodoDatabase = TodoDatabaseImpl()
}
```

arkivanov commented:
> Also I can't compile the Swift project because of strange errors:
> ![image](https://user-images.githubusercontent.com/26204457/79616022-f263e100-80fb-11ea-8012-7c212d0b0fe8.png)
> ![image](https://user-images.githubusercontent.com/26204457/79616087-10314600-80fc-11ea-9daf-e898ab3274e0.png)
> 
> For some reason it does not like passing `@Published` and `@ObservedObject` objects to Kotlin classes. But if I e.g. specify types explicitly, then it works:
> 
> ```swift
> class ControllerDeps: ObservableObject {
>     @Published var storeFactory: StoreFactory = LoggingStoreFactory(
>         delegate: TimeTravelStoreFactory(fallback: DefaultStoreFactory()),
>         logger: DefaultLogger(),
>         mode: LoggingMode.full,
>         eventTypes: StoreEventType.Companion().ALL
>     )
>     
>     @Published var database: TodoDatabase = TodoDatabaseImpl()
> }
> ```

Perhaps because my Xcode version is not very fresh, but I can't update it at the moment.

sgallese commented:
Odd, worked for me locally prior.

To test, I updated to latest Xcode 11.4.1 (11e5003a)
Cloned project fresh and switched to PR branch
run `./gradlew :sample:todo-darwin-umbrella:iosX64MainBinaries` (it was surprisingly fast, less than 2 minutes!)
Open project in Xcode, build succeeds, able to deploy to simulator.

sgallese commented on sample/todo-app-ios/todo-app-ios/list/TodoList.swift:
```diff
@@ -0,0 +1,36 @@
+    
+    var body: some View {
+        List(listView.model?.items ?? []) { item in
+            NavigationLink( destination: TodoDetails(id: item.id)) {
```
0cd0fe

arkivanov commented on sample/todo-app-ios/todo-app-ios/list/TodoRow.swift:
```diff
@@ -13,12 +13,11 @@ struct TodoRow: View {
     
     var body: some View {
-        Button(action: self.onItemClicked) {
+        Button(action: {}) {
```
Do we need this to be Button now? Maybe just text?

sgallese replied:
Removed the this Button, using only HStack as parent view now.
Oddly, the Trash Button no longer registers clicks. The NavigationLink is getting a higher priority.
Changing the Trash Button to simply an Image with onTapGesture resolves.
2f6d582

The click radius of the Trash button can be confusingly close to the chevron though.
We may want to have a more typical "iOS" delete behavior (usually a swipe).
I implemented that here, but we can revert if you don't like.
9f085ba

<img width="230" alt="Screen Shot 2020-04-23 at 2 33 11 PM" src="https://user-images.githubusercontent.com/140911/80152141-ffefef80-856f-11ea-86ba-931a22ecd609.png">

arkivanov commented on sample/todo-app-ios/todo-app-ios/list/TodoList.swift:
```diff
@@ -15,7 +15,7 @@ struct TodoList: View {
     var body: some View {
         List(listView.model?.items ?? []) { item in
-            NavigationLink( destination: TodoDetails(id: item.id)) {
+            NavigationLink( destination: TodoDetailsParent(id: item.id)) {
```
I suspect that `TodoDetailsParent` is created for every item in the list. Could you please check? Shouldn't there be a callback for navigation?

sgallese replied:
Yes, TodoDetailsParent is initialized for every item instance.
Fortunately, the majority of logic for TodoDetailsParent resides in body/appear/disappear, which will not be called until the view actually appears.

Unfortunately, yes, NavigationLink's standard behavior is to initialize the View struct for each row item, even if the user hasn't clicked on that item.
Since no complex logic is performed in the TodoDetailsParent initializer, I think we're OK for now.
If we did have complex logic, some developers approach the problem by hacking together a lazy view (e.g., https://stackoverflow.com/q/57594159).

arkivanov commented on sample/todo-app-ios/todo-app-ios/details/TodoDetailsParent.swift:
```diff
@@ -0,0 +1,48 @@
+import SwiftUI
+import TodoLib
+
+struct TodoDetailsParent: View {
```
Can we use this approach (Parent, Lifecycle, etc.) for List screen?

sgallese replied:
Of course :) !

Well, I thought this would be easy... but ran into some problems while testing.
Here is the code which essentially copies TodoDetailsParent but for TodoListParent:
33081ac

However, this code throws this error if you go from the TodoListParent -> TodoDetailsParent -> TodoListParent:
`Uncaught Kotlin exception: kotlin.IllegalStateException: Expected lifecycle state INITIALIZED, actual DESTROYED`

This means that if I call lifecycle.onDestroy(), I assume I should never call lifecycle.onCreate() on that same lifecycle.

Another approach could be to create the lifecycle on view initializer, setup the controller in view body, and then have view appear map to start/resume and disappear to pause/stop. I don't know how I'd map destroy though.
9d5d29

The problem with this code, and I'm not sure why, is I get a duplicate ListStore error.
```
[MVIKotlin] ListStore: created
[MVIKotlin] ListStore (RESULT, Unit): kotlin.Unit
[MVIKotlin] TodoAddStore: created
[MVIKotlin] ListStore: created
Instances of kotlin.Error, kotlin.RuntimeException and subclasses aren't propagated from Kotlin to Objective-C/Swift.
Other exceptions can be propagated as NSError if method has or inherits @Throws annotation.
Uncaught Kotlin exception: kotlin.IllegalStateException: Duplicate store: ListStore
```

And then my last idea is, well, why not just create a new lifecycle each time the view appears:
11d62ad

Unfortunately, that causes a very off visual issue when navigating from list view to detail view then back to list view:
![final_5ea230f48c77bb0013f7247d_736304](https://user-images.githubusercontent.com/140911/80162222-911e9080-8587-11ea-854c-5ef93e92f413.gif)

So, I simply reset the project back to where it was for now.
I can spend some more time thinking about this... but not sure if I have an easy solution.

Full exploration of problem:
https://gitduck.com/watch/5ea21e5055c63aa9c4ac1d38

arkivanov commented:
@sgallese What do you think about my comments above?

sgallese commented:
> @sgallese What do you think about my comments above?

Doh! Sorry I didn't see. Looking at them now, they look doable, will tackle shortly.

arkivanov commented:
No rush! Just a friendly reminder!

sgallese commented:
I wonder why CI cancelled the iOS job :( I'll commit a blank change to rerun

arkivanov commented:
I'm not sure but the `SceneDelegate` still manages the `TodoListReaktiveController`. And the `TodoDetailsReaktiveController` is managed by `TodoDetailsParent` view. In the first case the controlled is destroyed in `sceneWillResignActive` callback, in the second case the controlled is destroyed in View's `onDisappear` callback. Why is it like this?

sgallese commented:
The list uses the scene lifecycle and the details uses the view lifecycle.
Yes, I agree, dealing with two separate lifecycles is probably confusing for users of the sample app.
I think we'd want both to use the view lifecycle if possible.

Will need some more time to think of an issue and play around with the issues I'm facing in https://github.com/arkivanov/MVIKotlin/pull/33#discussion_r414212391

arkivanov commented:
Ok! Please re-request the review when there is something to check :-)

arkivanov commented:
> dealing with two separate lifecycles is probably confusing for users

I believe lifecycles should be separate since they are different screens.

Also ideally we should some how keep a controller alive (not destroyed) while it is inactive (stopped or its view is destroyed). The idea is when you are navigating from screen A to a screen B then A's view might be destroyed but the controlled may stay alive. And later when you return to screen A, its view is just recreated. This is how it works on Android. Not sure if it is possible with SwiftUI, we should check.

sgallese commented:
Ah, good point.

Sorry, I didn’t word clearly- yes, two separate life cycles is appropriate, but having one life cycle in the scene vs one life cycle in the view might be confusing. I can spend a bit more time making sure both life cycles are wired up in views.

Hopefully keeping the controller in the view body allows it to survive the view lifecycle. I’ll take another look at the Android sample app to see if I have any other ideas.

arkivanov commented:
Ok, let me know if you need any help. We can merge the PR and then I could also check what's going on. But better to make it as clean as possible before merging. Thanks.

sgallese commented:
I think I have a slightly better solution, but I think I need help for `LifecycleRegistry`

My main question is- how should I think about lifecycle in relation to the controller and view lifecycle?

I have the parent list view create an instance of lifecycle and controller.
Then I call onViewCreated using the lifecycle.
Then I call lifecycle onCreate...onResume.
Now I navigate to another view, so I call lifecycle onPause...onDestroy

Now, if I return to the parent list view.
I try to reuse the same lifecycle and controller.
Then I call onViewCreated using lifecycle.
And then when I call lifecycle onCreate, I get this error
`Expected lifecycle state INITIALIZED, actual DESTROYED`

Code: 714daff

I then tried is to ensure that if I call lifecycle onDestroy,
that I create a new lifecycle the next time I enter the view.
However, if I do this, if I re-enter the view, none of the functionality works.
I assume this is because the controller was initialized to work with the previous lifecycle.

Code: c66f72f

On the Android fragments, the controller is created once.
Each time you enter the screen, it goes through the typical cycle:
onViewCreated
onCreate
...
onDestroy
Android app is using androidx-lifecycle-interop and not LifeCycleRegistry
It looks like whenever the Android app calls onViewCreated, viewLifecycleOwner.lifecycle returns a different instance of of Lifecycle.

arkivanov commented:
@sgallese Good questions! `Lifecycle` object can not be reused once destroyed (its `onDestroy` method is called). Same applicable for the Androidx Lifecycle. Android Fragments have two Lifecycles: one is Fragment instance's `Lifecycle`, another one is View instance's `Lifecycle`. Fragment views can be destroyed and then recreated, a new `Lifecycle` object will be associated with the new view. However Fragment may be same instance, so its `Lifecycle` object is still used. This is what happening when you push a fragment to the back stack.

I was always thinking that in iOS UIKit view and its ViewController both have same lifecycle, so view controller is destroyed when its view is destroyed. And when you push another ViewController previous goes to back stack but it's not destroyed (nor its view). But I'm not sure about SwiftUI. **If** in SwiftUI there is such an entity (like ViewController or Fragment) that has wider lifecycle **and** you want to retain MVI controller while view is destroyed **then** you need two separate `Lifecycles` via two separate instances of `LifecycleRegistry`.

Can we (for simplicity) retain List screen view when navigating to Details screen?

sgallese commented:
@arkivanov wow, this comment was very helpful.
I played around with having two lifecycles to mimic Android: one for the parent and one for the view itself. And it worked!
I was able to simplify TodoListParent further into one lifecycle.

5e2e70d

This approach works well if there is only one instance of a view though.
For example, I tried implementing this same strategy for TodoDetailsParent.
But you soon encounter problems because there are multiple instances of TodoDetailsParent
created via a list.
If you call onStop (but not onDestroy) for an instance of TodoDetailsParent, if you navigate to  another instance of TodoDetailsParent, when creating a new controller I encounter the error:
`kotlin.IllegalStateException: Duplicate store: TodoEditStore`

sgallese commented:
@arkivanov wow, this comment was very helpful.
I played around with having two lifecycles to mimic Android: one for the parent and one for the view itself. And it worked!
I was able to simplify TodoListParent further into one lifecycle (ViewLifecycle.swift).

e4eb1b1

One issue I found: this approach works well if there is only one instance of a view.
For example, I tried implementing this same strategy for TodoDetailsParent.
But you soon encounter problems because there are multiple instances of TodoDetailsParent
created via a list.
If you call onStop (but not onDestroy) for one instance of TodoDetailsParent, if you navigate to  another instance of TodoDetailsParent, when creating a new controller I encounter the error:
`kotlin.IllegalStateException: Duplicate store: TodoEditStore`
I instead need to ensure that onDestory is called prior to creating a new instance of TodoDetailsParent.

arkivanov commented:
@sgallese Good news! Regarding the issue: if you navigate back from the Details screen, its view and its controller should be destroyed. And then when you navigate to the Details again then a new instance should be created (all of them: controller, view, lifecycles).

sgallese commented:
Sweet, I think the current PR meets that specification. See if you think it’s ready for merge or needs any other changes. 

Definitely a slow process but I’ve been learning a lot about SwiftUI and this framework, so pretty happy.

arkivanov commented:
If you will make the sample app working without errors I could merge it. Then we can tweak it as needed.

arkivanov commented:
Cool! I will check how it works today and probably merge. There is also chance I won't be able to build it. Previously it was failing.
